pkg/root/trusted_root_create.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2023 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package root
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"time"
+
+	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+	prototrustroot "github.com/sigstore/protobuf-specs/gen/pb-go/trustroot/v1"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	FulcioTarget = "Fulcio"
+	RekorTarget  = "Rekor"
+	CTFETarget   = "CTFE"
+	TSATarget    = "TSA"
+)
+
+type RawTrustedRootTarget interface {
+	GetType() string
+	GetBytes() []byte
+}
+
+type BaseTrustedRootTarget struct {
+	Type  string
+	Bytes []byte
+}
+
+func (b *BaseTrustedRootTarget) GetType() string {
+	return b.Type
+}
+
+func (b *BaseTrustedRootTarget) GetBytes() []byte {
+	return b.Bytes
+}
+
+// NewTrustedRootFromTargets initializes a TrustedRoot object from a mediaType string and
+// a slice of targets. These targets are expected to be PEM-encoded public keys/certificate chains.
+// This method of constructing the TrustedRoot has some shortcomings which can be aided by manually
```

Comment:
If we don't address these, the concern would be that one would construct a trust root with NewTrustedRootProtobuf(json), then marshal it with `MarshalJSON`, and get back a different trust root. 

We should have a test that demonstrates JSON-input -> `TrustedRoot` struct -> JSON-output, where JSON-input == JSON-output.

What if the API takes in a `TrustedRoot` struct, fully populated, and outputs JSON? And if certain fields are missing, then we can have reasonable defaults (e.g if `TransparencyLog.HashFunc` is not set, then pick `SHA256`)

Reply:
So there's couple different things to note here:
* The only thing that this new method does is it constructs a `TrustedRoot` from given targets; this does not concern the JSON representation of the `TrustedRoot`. On the contrary this is most useful when there is no `TrustedRoot` yet and one needs to be constructed from public keys/certs.
* However, I understand your concern about having the ability to marshal `TrustedRoot` - which was also added by this PR - I think the only problem right now is the "Uri" argument which I left unfilled for `CertificateAuthority` struct, so I probably need to fix that and I can also add a test that you suggest.
* I think taking in a `TrustedRoot` struct fully populated defeats the purpose of this PR - that won't solve a lot of what I need in https://github.com/sigstore/scaffolding/pull/1191 and most of the code will need to stay there to create the `TrustedRoot` struct in the first place.

So what I propose is adding the test as you suggested, and fixing any issues that pop up, but otherwise leaving this PR as is. Does that sound good?